### ISSUE TITLE
Enable the use of local record cleaner rules

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#1403]
The request is for a verifier to be able to upload a set of locally relevant rules that can be used to filter records, e.g. where local phenology is different to the phenology defined nationally.

@johnvanbreda has suggested 3 possibilities:
1. Modify the existing time of year rules so that they are linked to a geographic region, e.g. England, or a VC boundary.
2. Allow the user to upload a set of rules in a spreadsheet. The rules are then parsed to create a long, complex Elasticsearch filter which returns the records that pass or fail as required.
3. Allow the user to upload a set of rules in a spreadsheet. Then, run a script within Elasticsearch that finds the records of each given taxon within the verifier’s VC boundary, and sets a custom flag in the ES data according to the pass/fail result. This can then be used in Elasticsearch filters, or added as a flag column to the data grid. This gives the closest end-result to simply allowing verifiers to define their own rules.
 
The above objects are in order of flexibility, with 1 being the least flexible and simplest to implement and 3 being the most flexible and hardest to implement. The main issue with both #2 and #3 will be the impact on Elasticsearch performance.

A benefit of options 2 and 3 is that they allow county-based verifiers to be independent and use local rules or not as they choose. 

Initial testing to be carried out on the dev server to further explore options 2 and 3 and to see what impact they might have on performance.

**Comment by johnvanbreda** [comment]
Coding for this done, will set up on the test environment...

**Comment by johnvanbreda** [comment]
This is now available to test on the main verification page on the test iRecord install.

**Comment by kitenetter** [comment]
@johnvanbreda great to see this. Have tried it out this evening.

1. Setting up new rules: Geographic limits
When choosing a named place, such as a vice-county, it would be good if the relevant boundary was displayed on the map.

2. Setting up the spreadsheet
I have a number of questions about the spreadsheet, probably easiest to talk these through over zoom.

3. Importing a spreadsheet
I set up a simple spreadsheet to list rare species, that looked like this:

![image](https://user-images.githubusercontent.com/16039111/227064496-5413973d-d7e5-4909-ac53-eaee8013a84b.png)

On trying to import it I got an error message:

![image](https://user-images.githubusercontent.com/16039111/227064538-cdb5b6f5-8539-4d11-96af-4edaa6c4bd34.png)

Why is it not recognising the taxon id and rule type columns?

**Comment by johnvanbreda** [comment]
@kitenetter I've implemented the location boundaries on the map for point 1. For point 3 can you send me the spreadsheet?

Would tomorrow be a good day for a Zoom?

**Comment by johnvanbreda** [comment]
* I've fixed the issues uploading the spreadsheet.
* Note that the rule type should now be "species recorded" not "species_recorded".
* The format for reverse rule is now "yes".
* Answering a question about how "limit to" fields interact when specified in both a rule and a ruleset. The answer is that both checks must pass. So, if a rule has limit to grid ref SY1234 and a ruleset has limit to grid ref SY2345 then the rule will never actually run.

**Comment by johnvanbreda** [comment]
@kitenetter whenever you are ready I can add the Custom Rules button to the live iRecord verificaion pages, assuming you are happy that it's been tested enough.

**Comment by kitenetter** [comment]
No, it needs testing and guidance. That is in progress but will probably take a while yet.

**Comment by kitenetter** [comment]
@johnvanbreda last time I looked I could no longer find the test implementation of this, is it still available?

**Comment by kitenetter** [comment]
Enable this on live site for further testing before making available more widely.

**Comment by johnvanbreda** [comment]
@kitenetter I've added the Custom Rules button to the main verification page https://irecord.org.uk/verification but have configured it to show only for users with admin rights. The iRecord support accounts should have admin permissions so is that enough for you to check it? I can remove the permission setting when ready.

**Comment by kitenetter** [comment]
Will add a new issue to cover testing. Closing this issue.
